# gallvp/prepngs: Output

## Introduction

This document describes the output produced by the pipeline. The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

- [BAM to FastQ](#bam-to-fastq) - Convert input Bam files to FastQ files
- [FastQC Raw](#fastqc-raw) - Raw read QC statistics
- [FASTP](#fastp) - Adapter trimming
- [FastQC Trimmed](#fastqc-trimmed) - QC statistics for the reads trimmed by FASTP
- [CAT](#cat) - Concatenate reads by group
- [Seqkit/fq2fa](#seqkitfq2fa) - Convert trimmed reads to Fasta
- [MultiQC](#multiqc) - Aggregate report describing results and QC from the whole pipeline
- [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution

### BAM to FastQ

The conversion is performed using the `bam2fastq` command of the [PBTK](https://github.com/PacificBiosciences/pbtk) toolkit from Pacific Biosciences.

### FastQC Raw

<details markdown="1">
<summary>Output files</summary>

- `fastqc_raw/`
  - `*_fastqc.html`: FastQC report containing quality metrics.
  - `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

</details>

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

### FASTP

<details markdown="1">
<summary>Output files</summary>

- `fastp/`
  - `fail`
    - `*.fail.fastq.gz`: Reads which failed to pass the filter
  - `html`
    - `*.fastp.html`: Sample wise HTML report
  - `json`
    - `*.fastp.json`: Sample wise JSON report
  - `log`
    - `*.fastp.log`: Sample wise log file
  - `pass`
    - `*.fastp.fastq.gz`: Reads which passed the filter

</details>

[FASTP](https://github.com/OpenGene/fastp) is an ultra-fast all-in-one FASTQ preprocessor to perform QC, adapter trimming, filtering, splitting and merging.

### FastQC Trimmed

<details markdown="1">
<summary>Output files</summary>

- `fastqc_trim/`
  - `*_fastqc.html`: FastQC report containing quality metrics.
  - `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

</details>

FastQC is applied to the trimmed reads from FASTP `*.fastp.fastq.gz`.

### CAT

<details markdown="1">
<summary>Output files</summary>

- `groups/`
  - `fastq`
    - `*.merged.fastq.gz`: Concatenated fastq file

</details>

Samples with the same `group` column in the `samplesheet.csv` are concatenated together. For single-end samples, a single `*.merged.fastq.gz` file is created. For paired-end samples, two separate files for `reads_1` and `reads_2` are saved. The concatenation is performed with the [cat](https://www.linfo.org/cat.html) command.

### Seqkit/fq2fa

<details markdown="1">
<summary>Output files</summary>

- `groups/`
  - `fasta`
    - `*.fa.gz`: Concatenated fasta file

</details>

Concatenated FastQ files `*.merged.fastq.gz` are converted into Fasta files with `fq2fa` command of the [Seqkit](https://bioinf.shenwei.me/seqkit/) toolkit.

### MultiQC

<details markdown="1">
<summary>Output files</summary>

- `multiqc/`
  - `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  - `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  - `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
